### ISSUE TITLE
My app starts before the restore is finished and I get isAuthenticated: false

### ISSUE THREAD

**Issue by ieugen** [jerelmiller/redux-simple-auth#20]
Hello,

I'm writing an app with create-react-app + react-router 4 + redux + redux-simple-auth. My problem is that when the application starts, it displays the Login page even though the restore operation is done successfully. 

I have some private routes [1] and a Login Route. 
When the application start it renders the routers and it gets redirected to a private route. 
Because the restore has not yet completed the state is "un-authenticated" so I get to see the login page. Just after that, the restore happens but I am already on the login page.

Is there a way to wait for the restore and then bootstrap my application ?

````
const Root = ( { store } ) => (
    <Provider store={store}>
          <Router>
              <div>
                <ul>
                  <li><Link to="/c/dashboard">Dashboard</Link></li>
                  <li><Link to="/c/topics">Topics</Link></li>
                  <li><Link to="/c/profile">Profile</Link></li>
                </ul>
                <hr/>
                <Switch>
                  <PrivateRoute path="/c/:filter?" component={SchedulingRouter} />
                  <Route path="/login" component={Login} />
                  <Redirect from='/' to="/c/dashboard" />
                  <Route component={NoMatch}/>
                </Switch>
              </div>
          </Router>   
    </Provider>
);
````

````
import React, { Component } from 'react';
import { connect } from 'react-redux'
import { authenticate } from 'redux-simple-auth'
import { FormattedMessage } from 'react-intl'

import { Redirect } from 'react-router'
import { getIsAuthenticated, getSessionData } from 'redux-simple-auth'

const mapStateToProps = (state, ownProps) => {
  return {
    isAuthenticated: getIsAuthenticated(state),
    session: getSessionData(state)
  }
}

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    onLogin: (username, password) => {
      dispatch(authenticate('credentials', {
        username: username,
        password: password
      }))
    }
  }
}

class Login extends Component {

  constructor(props) {
    super(props)
    this.state = {
      username: '',
      password: '',
      fireRedirect: false
    }
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  handleLogin = (event) => {
    event.preventDefault();
    const {username, password } = this.state
    this.props.onLogin( username, password )
    this.setState({fireRedirect: true})
  }

  handleInputChange(event) {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  }

  render() {
    const { from } = this.props.location.state || { from: { pathname: '/' } }
    const { fireRedirect } = this.state
    return (
      <div>
        <div className="center-block loginContainer">
          <form className="form form--auth" method="POST" onSubmit={ (e) => this.handleLogin(e) }>
            <p>error message here</p>
            <div className="form-group">
              <input className="form-control" id="form-login-email"
                name="username"
                value={this.state.username}
                onChange={this.handleInputChange}
              />
            </div>
            <div className="form-group">
              <input className="form-control" type="password" id="form-login-pass"
                name="password"
                value={this.state.password}
                onChange={this.handleInputChange}
              />
            </div>

            <div className="form-group text-center actions">
              <button className="btn btn-wide btn-primary" type="submit" >
                  Login
              </button>
            </div>
          </form>
          {fireRedirect && (
          <Redirect to={from || '/'}/>
          )}
        </div>
      </div>
    );
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(Login);
````

[1] https://reacttraining.com/react-router/web/example/auth-workflow

**Comment by ieugen** [comment]
I managed to fix it by adding some state to monitor if the application has started:

````
import React from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux'

import Loading from '../components/Loading'

const mapStateToProps = (state, ownProps) => {
  return {
    applicationHasStarted: state.auth.hasRun
  }
}

/**
* Wait the application to startup ( authenticationHasRun = true )
* Display a Loading page until startup is done.
*/
const WaitApplicationStartup = ({ component: Component, applicationHasStarted, ...rest }) => (
  <Route {...rest} render={ props => (
    applicationHasStarted ? ( <Component {...props}/> ) : ( <Loading /> )
  )}/>
)
export default connect(mapStateToProps)(WaitApplicationStartup);
````

And use it: 

````
import App from '../containers/App';
import WaitApplicationStartup from '../containers/WaitApplicationStartup';

const Root = ( { store } ) => (
    <Provider store={store}>
        <Router>
          <WaitApplicationStartup component={App} />
        </Router>
    </Provider>
);
````

**Comment by jerelmiller** [comment]
Hey @ieugen, thanks for the inquiry!

There is indeed a bit easier way to know when the state has been restored. When the store first boots up, it will ask your authenticator's `restore` function to validate the authenticated data. Assuming that is successful, this will dispatch a restore action. Your best bet might be to listen for that action and toggle some state in your redux store to let the app know the session state has been initialized.

The action types are currently exported so you should be able to get access to this easily!

```javascript
import { actionTypes } from 'redux-simple-auth'

// will be available as actionTypes.RESTORE
```

Hopefully this helps! I think you have the right idea to have some state represent whether the session has been restored or not. Hopefully this action helps with this.

This project is still very young and I'd love to explore more of an integration with react-router. Its on my TODO list and hopefully we can start exploring soon!

If you have any suggestions about the process, definitely let me know! Thanks again!

**Comment by ieugen** [comment]
Hi, 
Thanks for the fast reply. That is exactly what I ended up doing, but forgot to add the example:
````
const initialState = {
  hasRun: false
}

const authReducer = (state = initialState , action) => {
  switch (action.type) {
    case actionTypes.RESTORE:
      return {
        hasRun: true
      }
    case actionTypes.RESTORE_FAILED:
        return { hasRun: true };
    default:
      return state;
  }
}
export default authReducer
````

I finally made the integration to work properly. A property that informs us if the session was restored would be awesome.

**Comment by jerelmiller** [comment]
Great suggestion! I'll try and incorporate something soon that does this very thing.

**Comment by jerelmiller** [comment]
It took me long enough, but I have published a version that has this behavior now built-in. You can now import the `getIsRestored` selector and select state based on that. It was published along with v0.4.0 so feel free to update if you want to use the built-in version.

**Comment by ieugen** [comment]
The guys from my team decided to go DYI. Not sure if they will change ti redux-simple-auth again.
